### PR TITLE
Fix #4855: Blacklist rcParams that aren't style

### LINKED ISSUE #4855: Limit what `style.use` can affect?
This might be a controversial proposal, so I thought I'd get some discussion first.

In `astropy`, there is a "house style", which, surprising to me, included `interactive: True`.  I wouldn't have expected a style to affect the interactive setting (which I see more as a user preference than affecting the style of the output).  I'm going to advocate that `astropy` removes that but relatedly...

Would it make sense to filter out rcParams that don't really relate to style in `style.use`?  I realize this somewhat breaks DWIM, but it does prevent a certain kind of shooting oneself in the foot...

### PR DESCRIPTION
**Pull request by mdboom**
~~Needs a what's new~~

### CODE REVIEW

File: lib/matplotlib/style/core.py [hunk 1]
```diff
@@ -34,11 +34,27 @@
...
+    return dict((k, v) for (k, v) in d.items() if k not in STYLE_BLACKLIST)
+
+
 def is_style_file(filename):
     """Return True if the filename looks like a style file."""
     return STYLE_FILE_PATTERN.match(filename) is not None
 
 
+def _apply_style(d):
+    mpl.rcParams.use(blacklist_style_params(d))
```
Comment (efiring) [inline lib/matplotlib/style/core.py]: Missing a leading underscore.

File: lib/matplotlib/style/core.py [hunk 2]
```diff
@@ -34,11 +34,27 @@
...
 
+# A list of rcParams that should not be applied from styles
+STYLE_BLACKLIST = set([
+    'interactive', 'backend', 'backend.qt4', 'webagg.port',
+    'webagg.port_retries', 'webagg.open_in_browser', 'backend_fallback',
+    'toolbar', 'timezone', 'datapath', 'figure.max_open_warning',
+    'savefig.directory', 'tk.window_focus', 'hardcopy.docstring'])
+
+
+def _blacklist_style_params(d):
```
Comment (efiring) [inline lib/matplotlib/style/core.py]: The name is not so clear in indicating that a dict of valid params will be returned; how about `_screened_style_params` or `_valid_style_params`?

File: lib/matplotlib/style/core.py [hunk 3]
```diff
@@ -34,11 +34,27 @@
...
+    return dict((k, v) for (k, v) in d.items() if k not in STYLE_BLACKLIST)
```
Comment (efiring) [inline lib/matplotlib/style/core.py]: Would it make sense to raise a warning if a a blacklisted key is found?

**Comment by tacaswell** [comment]
Should also remove all of the black-listed entries from `classic`

**Comment by mdboom** [comment]
@tacaswell: Good point about removing blacklisted entries from the classic style.  Done.

**Comment by mdboom** [comment]
I think this one is good to go.

File: lib/matplotlib/style/core.py [hunk 4]
```diff
@@ -71,18 +96,18 @@ def use(style):
 
     for style in styles:
         if not cbook.is_string_like(style):
-            mpl.rcParams.update(style)
+            _apply_style(style)
             continue
         elif style == 'default':
             mpl.rcdefaults()
```
Comment (WeatherGod) [inline lib/matplotlib/style/core.py]: There is a bit of a disconnect here. rcdefaults() is more than just style stuff, but it doesn't make sense to blacklist entries there, either, does it?

File: lib/matplotlib/style/core.py [hunk 5]
```diff
@@ -34,11 +35,35 @@
 STYLE_FILE_PATTERN = re.compile('([\S]+).%s$' % STYLE_EXTENSION)
 
 
+# A list of rcParams that should not be applied from styles
+STYLE_BLACKLIST = set([
+    'interactive', 'backend', 'backend.qt4', 'webagg.port',
+    'webagg.port_retries', 'webagg.open_in_browser', 'backend_fallback',
+    'toolbar', 'timezone', 'datapath', 'figure.max_open_warning',
+    'savefig.directory', 'tk.window_focus', 'hardcopy.docstring'])
```
Comment (WeatherGod) [inline lib/matplotlib/style/core.py]: what about the default keymappings?

**Comment by WeatherGod** [comment]
I think this raises a bit more general question of what is the line between styles and settings. There are tons of other settings such as the default keymappings, animation settings, and such that may or may not make sense as a "style". I am fine with what has been chosen so far, it is just a question of how much more do we want to blacklist.

**Comment by WeatherGod** [comment]
Also, do we know if people have been using style.use() as a convenient way to apply custom rcParams? If so, then this change could break that use-case (not that it would have been a good thing to do in the first place, though)

File: lib/matplotlib/style/core.py (hunk 5 above)
Comment (mdboom) [inline lib/matplotlib/style/core.py]: In that case, I thought using style might be handy, e.g. an Emacs key style (not that one could do that given our current framework, but you get the idea).

**Comment by mdboom** [comment]
> Also, do we know if people have been using style.use() as a convenient way to apply custom rcParams? 

No, but I know of at least three cases where style was used to turn interactive on and caused end users no end of trouble... ;)

**Comment by WeatherGod** [comment]
This is all fine, but we still don't have a definition that can help guide us in choosing what should and shouldn't be allowed in a style file. So, to get the conversation rolling, I'll propose a definition: "a non-style parameter is one whose change would have no effect after importing pyplot". This definition might not be good enough. Should the line be drawn at figure creation? axes creation?

**Comment by mdboom** [comment]
My definition is things that affect the appearance of the plot are style.  Things that are related to the environment/platform in use and other technical considerations are not.

**Comment by mdboom** [comment]
@tacaswell: Any thoughts on the definition of "style parameters" here?

**Comment by tacaswell** [comment]
I would say it has to have a visual effect, rather than a ui effect (interactive, key bindings, default file types) which i think lines up with what @mdboom said and is a super set of @weathergod.  Things like dpi are on the edge and should probably be allowed.

**Comment by mdboom** [comment]
I've added a comment about what should be blacklisted.  I hope that addresses the comments above, and otherwise this is good-to-merge.

**Comment by mdboom** [comment]
I think this is ready-to-merge, unless there are other comments.

File: lib/matplotlib/style/core.py (hunk 4 above)
Comment (efiring) [inline lib/matplotlib/style/core.py]: @mdboom, I think @weathergod has a good point here: 'default' style should modify only non-blacklisted things, shouldn't it?

**Comment by efiring** [comment]
@mdboom, see @weathergod's first inline comment.

File: lib/matplotlib/style/core.py (hunk 4 above)
Comment (mdboom) [inline lib/matplotlib/style/core.py]: I see.  That makes sense.

Comment (mdboom) [inline lib/matplotlib/style/core.py]: Fixed.

**Comment by efiring** [comment]
I have the impression that some might want tighter restrictions--a larger blacklist.  Merging this does not preclude such changes, but it seems worthwhile to get the mechanism in place without further delay.

**Comment by jenshnielsen** [comment]
Should this be backported to 2.0?

**Comment by mdboom** [comment]
yes.

**Pull request merged**